[flight recorder] record process group configuration

shengbao-zheng opened the pull request:
Summary: Record process group configuration (i.e. ranks involved in a process group) to facilitate NCCL related debugging.

Differential Revision: D53792087




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1500,6 +1501,15 @@ const int& ProcessGroupNCCL::globalRank() const {
+const std::vector<uint64_t>& ProcessGroupNCCL::groupRanks() const {
+  if (options_->global_ranks_in_group.empty()) {
+    static std::vector<uint64_t> globalRanks(size_);
+    std::iota(globalRanks.begin(), globalRanks.end(), 0);
```
Before this, can we verify this is the default PG by checking if uid_ == 0?

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -724,6 +724,9 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // return the rank_ of the the very first PG created, aka, default global PG.
   const int& globalRank() const;
 
+  // Returns the global ranks of of a PG.
```
nit, 2 'of'

shuqiangzhang commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -475,6 +477,18 @@ struct NCCLTraceBuffer {
+      return;
+    }
+    std::lock_guard<std::mutex> guard(mutex_);
+    if (pg_id_to_ranks_.find(pg_id) != pg_id_to_ranks_.end()) {
```
When will this happen? or we just need to update the map whenever we call record_pg by pg_id_to_ranks_[pg_id] = ranks;

shuqiangzhang commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -475,6 +477,18 @@ struct NCCLTraceBuffer {
     return id_++;
   }
 
+  void record_pg(size_t pg_id, std::vector<uint64_t> ranks) {
```
nit: more specific, record_pg_ranks?

shengbao-zheng replied:
So far I think we only need the ranks involved in a PG so we can use record_pg_ranks for now. If more PG info is needed in the future, we can extend it and rethink the new name.

wconstab replied:
Does it make sense to include the nccl comm id in this too?

shengbao-zheng replied:
I think we can revisit this (adding nccl comm id) if the current info is not enough?

shuqiangzhang commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -660,10 +675,19 @@ struct NCCLTraceBuffer {
+    auto pg_config = new_dict();
+    for (auto it = pg_id_to_ranks_.begin(); it != pg_id_to_ranks_.end(); it++) {
+      auto ranks = new_list();
+      for (uint64_t rank : it->second) {
```
nit, const auto&

shuqiangzhang commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -660,10 +675,19 @@ struct NCCLTraceBuffer {
       entries.push_back(dict);
     }
+    auto pg_config = new_dict();
+    for (auto it = pg_id_to_ranks_.begin(); it != pg_id_to_ranks_.end(); it++) {
```
nit: `for (const auto& [pg_id, ranks] :  pg_id_to_ranks)`

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -573,6 +587,7 @@ struct NCCLTraceBuffer {
     // Update whenever changing contents or formatting of the dump
     // (minor when adding fields, major when changing existing fields)
     c10::IValue version_val = "1.1";
```
It should be 1.2 after adding the field

shengbao-zheng committed:
[flight recorder] record process group configuration (#120262)

Summary:

Record process group configuration (i.e. ranks involved in a process group) to facilitate NCCL related debugging.

Differential Revision: D53792087
